src/ucs/sys/event_set.c

```diff
@@ -0,0 +1,170 @@
+
+    event_set = ucs_malloc(sizeof(ucs_sys_event_set_t), "ucs_sys_event_set");
+    if (event_set == NULL) {
+        status = UCS_ERR_NO_MEMORY;
+        goto out_create;
+    }
+
+    /* Create epoll set the thread will wait on */
+    event_set->epfd = epoll_create(1);
+    if (event_set->epfd < 0) {
+        ucs_error("epoll_create() failed: %m");
+        status = UCS_ERR_IO_ERROR;
+        goto err_free;
+    }
+
+    *event_set_p = event_set;
+    return UCS_OK;
+
+err_free:
+    ucs_free(event_set);
+out_create:
+    return status;
+}
+
+ucs_status_t ucs_event_set_add(ucs_sys_event_set_t *event_set, int event_fd,
+                               ucs_event_set_type_t events)
+{
+    struct epoll_event raw_event;
+    int ret;
+
+    memset(&raw_event, 0, sizeof(raw_event));
+    raw_event.events   = ucs_event_set_map_to_raw_events(events);
+    raw_event.data.fd  = event_fd;
+
+    ret = epoll_ctl(event_set->epfd, EPOLL_CTL_ADD, event_fd, &raw_event);
+    if (ret < 0) {
+        ucs_error("epoll_ctl(epfd=%d, ADD, fd=%d) failed: %m", event_set->epfd,
+                  event_fd);
+        return UCS_ERR_IO_ERROR;
+    }
+
+    return UCS_OK;
+}
+
+ucs_status_t ucs_event_set_mod(ucs_sys_event_set_t *event_set, int event_fd,
+                               ucs_event_set_type_t events)
+{
+    struct epoll_event raw_event;
+    int ret;
+
+    memset(&raw_event, 0, sizeof(raw_event));
+    raw_event.events   = ucs_event_set_map_to_raw_events(events);
+    raw_event.data.fd  = event_fd;
+
+    ret = epoll_ctl(event_set->epfd, EPOLL_CTL_MOD, event_fd, &raw_event);
+    if (ret < 0) {
+        ucs_error("epoll_ctl(epfd=%d, MOD, fd=%d) failed: %m", event_set->epfd,
+                  event_fd);
+        return UCS_ERR_IO_ERROR;
+    }
+
+    return UCS_OK;
+}
+
+ucs_status_t ucs_event_set_del(ucs_sys_event_set_t *event_set, int event_fd)
+{
+    int ret;
+
+    ret = epoll_ctl(event_set->epfd, EPOLL_CTL_DEL, event_fd, NULL);
+    if (ret < 0) {
+        ucs_error("epoll_ctl(epfd=%d, DEL, fd=%d) failed: %m", event_set->epfd,
+                  event_fd);
+        return UCS_ERR_IO_ERROR;
+    }
+
+    return UCS_OK;
+}
+
+ucs_status_t ucs_event_set_wait(ucs_sys_event_set_t *event_set, int timeout_ms,
+                                event_set_handler_t event_set_handler)
+{
+    int nready;
+    struct epoll_event ep_events[UCS_EVENT_EPOLL_MAX_EVENTS];
+    ucs_sys_event_set_data_t event_data[UCS_EVENT_EPOLL_MAX_EVENTS];
+    int i;
+
+    if (event_set_handler == NULL) {
+        return UCS_ERR_INVALID_PARAM;
+    }
+
+    nready = epoll_wait(event_set->epfd, ep_events, UCS_EVENT_EPOLL_MAX_EVENTS,
+                        timeout_ms);
+    if ((nready < 0) && (errno != EINTR)) {
+        ucs_error("epoll_wait() failed: %m");
+        return UCS_ERR_IO_ERROR;
+    }
+    ucs_trace_data("epoll_wait(epfd=%d, timeout=%d) returned %d",
+                   event_set->epfd, timeout_ms, nready);
+
+    for (i=0; i < nready; i++) {
```

Comment:
we can call the callback "nready" times, each time with one event, it would be simpler API

Reply:
I thought this way at once.
But I worried that if epoll_wait returns `nready == 2` and first callback return `UCS_OK` second returns `UCS_ERR_XXX`, what status should I return?
That's why I created previous functions.
What do you think about this implementation?

Reply:
i think the callback can just return void

Reply:
Changed return void instead.